compiler/rustc_hir_analysis/src/coherence/inherent_impls.rs

```diff
@@ -143,7 +143,10 @@ impl<'tcx> InherentCollect<'tcx> {
 
         let id = id.owner_id.def_id;
         let item_span = self.tcx.def_span(id);
-        let self_ty = self.tcx.type_of(id).instantiate_identity();
+        let mut self_ty = self.tcx.type_of(id).instantiate_identity();
+        while let ty::Pat(base, _) = *self_ty.kind() {
+            self_ty = base;
+        }
```

Comment:
I explained it in a comment now. The comment is

> We allow impls on pattern types exactly when we allow impls on the base type.

Inherent impls are convenient for the unstable implementation, as I can just provide an `as_u32` method on `pat!(u32 as 1..)` instead of having to have a free method

Reply:
This is definitely worth a FIXME to remove this peeling and replace it with meaningful coherence rules (if not limited, e.g., to libcore or a specific `rustc_attr`)